pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -544,24 +550,25 @@ public final synchronized void readEntriesComplete(List<Entry> entries, Object c
         if (serviceConfig.isDispatcherDispatchMessagesInSubscriptionThread()) {
             // setting sendInProgress here, because sendMessagesToConsumers will be executed
             // in a separate thread, and we want to prevent more reads
-            sendInProgress = true;
-            dispatchMessagesThread.execute(safeRun(() -> sendMessagesToConsumers(readType, entries)));
+            dispatchMessagesThread.execute(safeRun(() -> {
+                if (sendMessagesToConsumers(readType, entries)) {
+                    readMoreEntries();
+                }
+            }));
```

Comment:
Note that this is not an equivalent change regarding synchronization. The `readMoreEntries` is no longer happening within the `sendMessagesToConsumers` lock.

Reply:
Both methods are synchronized themselves, but now there is a race where another thread could acquire the lock in between the two method calls. (I haven't reviewed all method calls to see if this is actually a problem or just a potential one. Given that we are using the `synchronized` keyword, I would guess this matters.

Reply:
After looking closer, I think it's acceptable to break up these two calls.

Reply:
I think it's ok here because `readMoreEntries ` is independent of `sendMessagesToConsumers`. and `readMoreEntries` also have `synchronized` key-word...

Reply:
Yes, I agree that the two are independent.